### PR TITLE
Update `crypto` in `@types/node` to use ArrayBufferView, add stream options to crypto stream types

### PR DESCRIPTION
**Pull request by ZaneHannanAU**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <https://nodejs.org/docs/latest/api/crypto.html>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by mohsen1 — APPROVED** [review]

File: types/node/index.d.ts [hunk 1]
```diff
@@ -5754,25 +5754,25 @@ declare module "crypto" {
...
-    export function createCipher(algorithm: string, password: any): Cipher;
-    export function createCipheriv(algorithm: string, key: any, iv: any): Cipher;
+    export function createCipher(algorithm: string, password: string | ArrayBufferView, options?: stream.TransformOptions): Cipher;
+    export function createCipheriv(algorithm: string, key: string | ArrayBufferView, iv: string | ArrayBufferView, options?: stream.TransformOptions): Cipher;
     export interface Cipher extends NodeJS.ReadWriteStream {
-        update(data: Buffer | DataView): Buffer;
-        update(data: string, input_encoding: Utf8AsciiBinaryEncoding): Buffer;
-        update(data: Buffer | DataView, input_encoding: any, output_encoding: HexBase64BinaryEncoding): string;
+        update(data: string | ArrayBufferView): Buffer;
+        update(data: string | ArrayBufferView, input_encoding: Utf8AsciiBinaryEncoding): Buffer;
```
Comment (Flarna) [inline types/node/index.d.ts]: This overload should not include `ArrayBufferView`.

File: types/node/index.d.ts [hunk 2]
```diff
@@ -5781,45 +5781,46 @@ declare module "crypto" {
...
     }
-    export function createVerify(algorith: string): Verify;
+    export function createVerify(algorith: string, options?: stream.WritableOptions): Verify;
     export interface Verify extends NodeJS.WritableStream {
-        update(data: string | Buffer | DataView): Verify;
-        update(data: string | Buffer | DataView, input_encoding: Utf8AsciiLatin1Encoding): Verify;
-        verify(object: string | Object, signature: Buffer | DataView): boolean;
+        update(data: string | ArrayBufferView): Verify;
+        update(data: string | ArrayBufferView, input_encoding: Utf8AsciiLatin1Encoding): Verify;
+        verify(object: string | Object, signature: Bufferalike): boolean;
```
Comment (Flarna) [inline types/node/index.d.ts]: Why is here `Bufferalike` and not `ArrayBufferView` used?

Comment (ZaneHannanAU) [reply types/node/index.d.ts]: @Flarna I missed a replacement value. Originally I added Bufferalike as just all of the ArrayBuffers that I knew of and I found out about ArrayBufferView after that …

File: types/node/index.d.ts [hunk 3]
```diff
@@ -5755,57 +5755,77 @@ declare module "crypto" {
...
         update(data: string | ArrayBufferView): Hmac;
-        update(data: string | ArrayBufferView, input_encoding: Utf8AsciiLatin1Encoding): Hmac;
+        update(data: string, input_encoding: Utf8AsciiLatin1Encoding): Hmac;
         digest(): Buffer;
         digest(encoding: HexBase64Latin1Encoding): string;
     }
-
+    export type CipherAADTypes = 'aes-128-ccm' | 'aes-192-ccm' | 'aes-256-ccm' | 'aes-128-gcm' | 'aes-192-gcm' | 'aes-256-gcm'
+    export type CipherAADOptions extends stream.TransformOptions {
+        authTagLength: number;
```
Comment (Flarna) [inline types/node/index.d.ts]: `authTagLength` is optional for gcm modes

File: types/node/index.d.ts [hunk 4]
```diff
@@ -5755,57 +5755,77 @@ declare module "crypto" {
...
+
     export function createCipheriv(algorithm: string, key: string | ArrayBufferView, iv: string | ArrayBufferView, options?: stream.TransformOptions): Cipher;
+    export function createCipheriv(algorithm: CipherAADTypes, key: string | ArrayBufferView, iv: string | ArrayBufferView, options: CipherAADOptions): CipherAAD;
+
     export interface Cipher extends NodeJS.ReadWriteStream {
         update(data: string | ArrayBufferView): Buffer;
-        update(data: string | ArrayBufferView, input_encoding: Utf8AsciiBinaryEncoding): Buffer;
-        update(data: ArrayBufferView, input_encoding: any, output_encoding: HexBase64BinaryEncoding): string;
+        update(data: string, input_encoding: Utf8AsciiBinaryEncoding): Buffer;
+        update(data: ArrayBufferView, input_encoding?: any, output_encoding: HexBase64BinaryEncoding): string;
```
Comment (Flarna) [inline types/node/index.d.ts]: I don't think an optional before a mandatory arg is allowed.

**Review by Flarna — COMMENTED** [review]
I think you have to merge master or rebase to get CI running.

File: types/node/index.d.ts (hunk 4 above)
Comment (Flarna) [reply types/node/index.d.ts]: CI will tell once you fix the merge conflict.

**Comment by ZaneHannanAU** [comment]
I have no idea what I'm doing but something might work

**Comment by ZaneHannanAU** [comment]
nope

**Comment by isLishude** [comment]
git pull --rebase

File: types/node/index.d.ts (hunk 4 above)
Comment (ZaneHannanAU) [reply types/node/index.d.ts]: yep

**Comment by ZaneHannanAU** [comment]
I hope travis will eventually do quick builds where it builds only the changed files and changed files dependants.

**Comment by ZaneHannanAU** [comment]
```
Error: Something went wrong in runWithListeningChildProcesses
    at fail (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:379:11)
    at ChildProcess.child.on (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:367:6)
    at emitNone (events.js:106:13)
    at ChildProcess.emit (events.js:208:7)
    at finish (internal/child_process.js:757:14)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! Test failed.  See above for more details.
```
... /cc @RyanCavanaugh

**Comment by isLishude** [comment]
You can test your code locally with `npm test` then push code here. @ZaneHannanAU

**Comment by ZaneHannanAU** [comment]
Looks like I'll need to. Travis keeps on exiting late anyway though… even if I test and pass it locally, will Travis actually pass it?

```travis-log
nodemailer-pickup-transport OK
npm OK
nsqjs OK
Error: Something went wrong in runWithListeningChildProcesses
    at fail (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:379:11)
    at ChildProcess.child.on (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:367:6)
    at emitNone (events.js:106:13)
    at ChildProcess.emit (events.js:208:7)
    at finish (internal/child_process.js:757:14)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! Test failed.  See above for more details.

travis_time:end:343b0f2c:start=1529621604625645302,finish=15296225438627
The command "npm test" exited with 1.

Done. Your build exited with 1.
```

**Comment by ZaneHannanAU** [comment]
though node did fail this time…

File: types/node/index.d.ts [hunk 5]
```diff
@@ -376,20 +376,27 @@ declare var Buffer: {
...
      * Creates a new Buffer containing the given JavaScript string {str}.
      * If provided, the {encoding} parameter identifies the character encoding.
      * If not provided, {encoding} defaults to 'utf8'.
      */
     from(str: string, encoding?: string): Buffer;
+    /**
+     * Creates a new Buffer using the passed {data}
+     * @param values to create a new Buffer
+     */
+    of(...items: number[]): Buffer;
```
Comment (Flarna) [inline types/node/index.d.ts]: `of` is not a member of `Buffer`, it's available in `Buffer` via base `Uint8Array` so it should not be here.

Comment (ZaneHannanAU) [reply types/node/index.d.ts]: Yeah; I just did a hotfix because a build failed because it couldn't find a definition of `Buffer.of`; so I added it.

**Comment by kjin** [comment]
This change seems to have broken our code, since we expected methods like `toString(encoding?: string)` to exist on the return value of `randomFillSync` and `randomBytes`.

I'm wondering what the motivation was behind this change... and how we can fix it (if it is something that needs to be fixed).

**Comment by Flarna** [comment]
@kjin `randomFillSync()` returns what it gets as first argument - if you pass a `DataView` then you get a `DataView` (which has no `toString(encoding)` not a `Buffer` so the change is not wrong.

But I think it could be improved by something like this: `export function randomFillSync<T extends ArrayBufferView>(buffer: T, offset?: number, size?: number): T` (similar for `randomFill`).

`randomBytes()` still returns a `Buffer` so this should be no issue.

**Comment by kjin** [comment]
@Flarna -- Thanks. I've opened a PR here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26886

**Pull request merged**